src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
+                        _logger.ResourceFilterShortCircuited((IFilterMetadata)state);
+
+                        var task = InvokeResultAsync(_resourceExecutingContext.Result);
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ResourceEnd;
+                            return task;
+                        }
+
+                        goto case State.ResourceEnd;
+                    }
+
+                case State.ResourceInside:
+                    {
+                        goto case State.ExceptionBegin;
+                    }
+
+                case State.ResourceEnd:
+                    {
+                        if (scope == Scope.Resource)
+                        {
+                            isCompleted = true;
+                            return TaskCache.CompletedTask;
+                        }
+
+                        Debug.Assert(scope == Scope.Invoker);
+                        Rethrow(_resourceExecutedContext);
+
+                        goto case State.InvokeEnd;
+                    }
+
+                case State.ExceptionBegin:
+                    {
+                        _cursor.Reset();
+                        goto case State.ExceptionNext;
+                    }
+
+                case State.ExceptionNext:
+                    {
+                        var current = _cursor.GetNextFilter<IExceptionFilter, IAsyncExceptionFilter>();
+                        if (current.FilterAsync != null)
+                        {
+                            state = current.FilterAsync;
+                            goto case State.ExceptionAsyncBegin;
+                        }
+                        else if (current.Filter != null)
+                        {
+                            state = current.Filter;
+                            goto case State.ExceptionSyncBegin;
+                        }
+                        else if (scope == Scope.Exception)
+                        {
+                            goto case State.ExceptionInside;
+                        }
+                        else
+                        {
+                            Debug.Assert(scope == Scope.Invoker || scope == Scope.Resource);
+                            goto case State.ActionBegin;
+                        }
+                    }
+
+                case State.ExceptionAsyncBegin:
+                    {
+                        var task = InvokeNextExceptionFilterAsync();
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ExceptionAsyncResume;
+                            return task;
+                        }
+
+                        goto case State.ExceptionAsyncResume;
+                    }
+
+                case State.ExceptionAsyncResume:
+                    {
```

Comment:
It would be nice to describe the possible scenarios related to exceptionContext here in a comment. exceptionContext == null -> no exception happened. exceptionContext != null && exceptionContext.Exception -> an exception happened but was handled. exceptionContext != null && exceptionContext.Exception == null -> An exception happened but was handled by an inner exception handler.

I'm making all of these kind of comments because with the current implementation of the invoker it's really hard to grasp these details.